cs4660/graph/graph.py

```diff
@@ -0,0 +1,278 @@
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash((self.from_node, self.to_node, self.weight))
+
+
+class AdjacencyList(object):
+    """
+    AdjacencyList is one of the graph representation which uses adjacency list to
+    store nodes and edges
+    """
+    def __init__(self):
+        # adjacencyList should be a dictonary of node to edges
+        self.adjacency_list = {}
+
+    # return True/False if node1 and node 2 are directly connected or not
+    def adjacent(self, node_1, node_2):
+        neighbors = self.neighbors(node_1)
+        for n in neighbors:
+            if n == node_2:
+                return True
+        return False
+    
+    # return all nodes which is adjacent from node
+    def neighbors(self, node):
+        list = []
+        for edge in self.adjacency_list[node]:
+            list.append(edge.to_node)     
+        return list
+
+    # return True if node does "NOT" exists otherwise return false
+    def add_node(self, node):
+        # node exists
+        if node in self.adjacency_list:
+            return False
+        # node not exists
+        self.adjacency_list[node] = []
+        return True
+        
+    # return true of node is removed otherwise return false
+    def remove_node(self, node):
+        if node not in self.adjacency_list:
+            return False
+        self.adjacency_list[node] = []
+
+        #remove edge which connected to node
+        for edges in self.adjacency_list.values():
+            for edge in edges:
+                if edge.to_node == node:
+                    self.remove_edge(edge)
+        return True
+
+    # add edge and return true otherwise, return false
+    def add_edge(self, edge):
+        if edge.from_node in self.adjacency_list and edge not in self.adjacency_list[edge.from_node]:
+            self.adjacency_list[edge.from_node].append(edge)
+            return True
+        return False
+
+    # remove edge and return true, otherwise, return false
+    def remove_edge(self, edge):
+        adjacent_edges = self.adjacency_list[edge.from_node]
+        if edge not in adjacent_edges:
+            return False
+        adjacent_edges.remove(edge)
+        return True
+
+class AdjacencyMatrix(object):
+    def __init__(self):
+        # adjacency_matrix should be a two dimensions array of numbers that
+        # represents how one node connects to another
+        self.adjacency_matrix = []
+        # in additional to the matrix, you will also need to store a list of Nodes
+        # as separate list of nodes
+        self.nodes = []
+
+    # return true if node_1 and node_2 is connected
+    def adjacent(self, node_1, node_2):
+        node_1_index =  self.__get_node_index(node_1)
+        node_2_index =  self.__get_node_index(node_2)
+        if self.adjacency_matrix[node_1_index][node_2_index] == None or node_1 not in self.nodes or node_2 not in self.nodes:
```

Comment:
this logic should check the other way, `node_1 not in self.nodes or node2 not in self.nodes or self.adjacency_matrix[node_1_index][node_2_index]` otherwise, if the node doesn't exist, you will still get index out of bound exception.